protocol/block/mining_api.go

```diff
@@ -0,0 +1,73 @@
+package block
+
+import (
+	"context"
+	"time"
+
+	"github.com/filecoin-project/go-filecoin/chain"
+	"github.com/filecoin-project/go-filecoin/mining"
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+// MiningAPI provides an interface to the block mining protocol.
+type MiningAPI struct {
+	addNewBlockFunc  func(context.Context, *types.Block) (err error)
+	chainReader      chain.ReadStore
+	mineDelay        time.Duration
+	startMiningFunc  func(context.Context) error
+	stopMiningFunc   func(context.Context)
+	createWorkerFunc func(ctx context.Context) (mining.Worker, error)
+}
+
+// New creates a new MiningAPI instance with the provided deps
+func New(
+	addNewBlockFunc func(context.Context, *types.Block) (err error),
+	chainReader chain.ReadStore,
+	blockMineDelay time.Duration,
+	startMiningFunc func(context.Context) error,
+	stopMiningfunc func(context.Context),
+	createWorkerFunc func(ctx context.Context) (mining.Worker, error),
+) MiningAPI {
+	return MiningAPI{
+		addNewBlockFunc:  addNewBlockFunc,
+		chainReader:      chainReader,
+		mineDelay:        blockMineDelay,
+		startMiningFunc:  startMiningFunc,
+		stopMiningFunc:   stopMiningfunc,
+		createWorkerFunc: createWorkerFunc,
+	}
+}
+
+// MiningOnce mines a single block in the given context, and returns the new block.
+func (a *MiningAPI) MiningOnce(ctx context.Context) (*types.Block, error) {
+	ts := a.chainReader.Head()
+
+	miningWorker, err := a.createWorkerFunc(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := mining.MineOnce(ctx, miningWorker, a.mineDelay, ts)
```

Comment:
MineOnce (formerly known as Once) was always creating its own worker; this at least puts such code in a single spot so we know it's done the same way here as in node.StartMining.